// discord-tictactoe/dist/src/bot/EventHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**@class Listens for events
*/
class EventHandler {
    constructor() {
        this.listeners = new Map();
        this.supportEvent('win');
        this.supportEvent('tie');
    }
    /**
     * @method registers eventname as a listener and processes it
     * @param {any} eventName The event to be listened for
     * @param {any} callback What to do when the event is heard.
     */
    registerListener(eventName, callback) {
        const array = this.listeners.get(eventName);
        if (array) {
            array.push(callback);
        }
        else {
            throw new Error(`Cannot register event "${eventName}" because it does not exist`);
        }
    }
    emitEvent(eventName, data) {
        var _a;
        (_a = this.listeners.get(eventName)) === null || _a === void 0 ? void 0 : _a.forEach(listener => listener(data));
    }
    supportEvent(eventName) {
        this.listeners.set(eventName, []);
    }
}
exports.default = EventHandler;
