// discord-rock-paper-scissor/src/index.js
const Discord = require('discord.js'),
    { getChoice, getChoices, getEmoji } = require('./utility'),
    color = require('./classes/color');

class rps {
    /**
     * 
     * @param {Object} options The options for rock paper scissors
     * @param {"dm" | "channel"} options.chooseIn Where should the bot message to ask for user's choice.
     * @param {String} options.readyMessage message to send while bot is waiting for users to choose their move.
     * @param {String} options.choiceTitle The title for embed sent in DM to get user's move.
     * @param {String} options.choiceDescription The description for embed sent in DM to get user's move.
     * @param {String} options.choiceReply The reply to the user after choosing process is finished.
     * @param {String} options.drawEndTitle The title for embed sent if game is ended as an draw.
     * @param {String} options.drawEndDescription The description for embed sent if game is ended as an draw.
     * @param {String} options.endTitle The title for embed sent if game is not an draw.
     * @param {String} options.endDescription The description for embed sent if game is not an draw.
     * @param {Colors} options.colors The colors for the embeds.
     */
    constructor(options = {}) {
        this.colors = new color(options.colors);
        this.chooseIn = options.chooseIn || "dm";
        this.readyMessage = options.readyMessage || "Choose your moves in DM";
        this.choiceTitle = options.choiceTitle || "Rock Paper Scissor";
        this.choiceDescription = options.choiceDescription || "Choose your move by clicking on the buttons";
        this.choiceReply = options.choiceReply || "Nerd chose {move}";
        this.drawEndTitle = options.drawEndTitle || "Game ended with an draw ðŸ˜ž";
        this.drawEndDescription = options.drawEndDescription || "{player1} chose : {player1move}\n\n{player2} chose : {player2move}";
        this.endTitle = options.endTitle || "Game ended victoriously for {winner}"
        this.endDescription = options.endDescription || "{winner} [ Winner ðŸ‘‘ ] chose : {winnermove}\n\n{looser} [ Looser ðŸ‘‘ ] chose : {loosermove}\n\n"

        if (this.chooseIn !== "dm" && this.chooseIn !== "channel") throw new Error("Choose in property should be either \"dm\" or \"channel\" but I got " + JSON.stringify(this.chooseIn));
    }

    /**
     * The solo mode for rock paper scissor. User VS Bot
     * @param {Discord.Message} message The messae object in which command was used
     * @param {Discord.Client} bot The client object
     */
    async solo(message, bot) {
        message.author = message.author || message.user;
        getChoice.bind(this)(message.author, message.channel).then(v => {

            const userChoice = v.choice;
            const sent = v.message;
            const choice = getEmoji(Math.floor(Math.random() * 3) + 1);

            let row = new Discord.MessageActionRow().addComponents(new Discord.MessageButton().setCustomId("y7ghjuiojioujoj").setDisabled(true).setStyle("SUCCESS").setEmoji("ðŸ•Š").setLabel("Game Ended"))

            if (userChoice === choice) { // draw
                sent.edit({ embeds: [{ color: this.colors.drawEmbed, title: this.drawEndTitle, description: this.drawEndDescription.replace(/{player1}/g, message.author.username).replace(/{player1move}/g, userChoice).replace(/{player2}/g, bot.user.username || "Bot").replace(/{player2move}/g, choice) }], components: [row] });
            } else if ((userChoice === "âœŠ" && choice === "âœŒï¸") || (userChoice === "ðŸ¤š" && choice === "âœŠ") || (userChoice === "âœŒï¸" && choice === "ðŸ¤š")) { // user win
                sent.edit({ embeds: [{ color: this.colors.endEmbed, title: this.endTitle.replace(/{winner}/g, message.author.username).replace(/{looser}/g, bot.user.username), description: this.endDescription.replace(/{winner}/g, message.author.username).replace(/{winnermove}/g, userChoice).replace(/{looser}/g, bot.user.username || "Bot").replace(/{loosermove}/g, choice) }], components: [row] });
            } else { // User loose
                sent.edit({ embeds: [{ color: this.colors.endEmbed, title: this.endTitle.replace(/{looser}/g, message.author.username).replace(/{winner}/g, bot.user.username), description: this.endDescription.replace(/{looser}/g, message.author.username).replace(/{loosermove}/g, userChoice).replace(/{winner}/g, bot.user.username || "Bot").replace(/{winnermove}/g, choice) }], components: [row] });
            }
        }).catch(e => {
            if (e.username) {
                if(!message.replied)message.reply(`I was unable to DM ${e.username}`);
                else message.followUp(`I was unable to DM ${e.username}`);
            } else {
                if(!message.replied)message.reply("There was a error in executing the command");
                else message.followUp("There was a error in executing the command");
                
                console.log(`[discord-rock-paper-scissor] : Error in Solo mode : `);
                console.log(e);
            }
        })
    }

    /**
     * The dyo mode for rock paper scissor. User VS Bot
     * @param {Discord.Message} message The messae object in which command was used
     * @param {Discord.User} player2 The Second Player's Discord User Object
     */
    async duo(message, player2) {
        message.author = message.author || message.user;
        if (!message || !message.author) throw new Error("Invalid Message Object");
        if (!player2 || !player2.username) throw new Error("Invalid Player 2 Object");

        const player1 = message.author;
        const sent = await message.channel.send({ embeds: [{ color: this.colors.readyEmbed, title: this.readyMessage }] });
        let no = false, player1Choice = "", player2Choice = "";

        if (this.chooseIn === "dm") {
            await message.channel.send({ content: `${player1.toString()}`, reply: { messageReference: sent.id } });
            await getChoice.bind(this)(player1, await player1.createDM()).then(v => player1Choice = v.choice).catch(e => no = e.username)
            await message.channel.send({ content: `${player2.toString()}`, reply: { messageReference: sent.id } });
            await getChoice.bind(this)(player2, await player2.createDM()).then(v => player2Choice = v.choice).catch(e => no = e.username);

            if (no !== false) return sent.edit({ components: [], embeds: [{ color: this.colors.errorEmbed, title: `I was unable to DM ${no}, so please open DM than try again.` }] })

            let row = new Discord.MessageActionRow().addComponents(new Discord.MessageButton().setCustomId("y7ghjuiojioujoj").setDisabled(true).setStyle("SUCCESS").setEmoji("ðŸ•Š").setLabel("Game Ended"))

            if (player1Choice === player2Choice) { // draw
                sent.edit({ embeds: [{ color: this.colors.drawEmbed, title: this.drawEndTitle, description: this.drawEndDescription.replace(/{player1}/g, message.author.username).replace(/{player1move}/g, player1Choice).replace(/{player2}/g, player2.username).replace(/{player2move}/g, player2Choice) }], components: [row] });
            } else if ((player1Choice === "âœŠ" && player2Choice === "âœŒï¸") || (player1Choice === "ðŸ¤š" && player2Choice === "âœŠ") || (player1Choice === "âœŒï¸" && player2Choice === "ðŸ¤š")) { // player 1 won
                sent.edit({ embeds: [{ color: this.colors.endEmbed, title: this.endTitle.replace(/{winner}/g, message.author.username).replace(/{looser}/g, player2.username), description: this.endDescription.replace(/{winner}/g, message.author.username).replace(/{winnermove}/g, player1Choice).replace(/{looser}/g, player2.username || "Bot").replace(/{loosermove}/g, player2Choice) }], components: [row] });
            } else { // player 2 won
                sent.edit({ embeds: [{ color: this.colors.endEmbed, title: this.endTitle.replace(/{looser}/g, message.author.username).replace(/{winner}/g, player2.username), description: this.endDescription.replace(/{looser}/g, message.author.username).replace(/{loosermove}/g, player1Choice).replace(/{winner}/g, player2.username || "Bot").replace(/{winnermove}/g, player2Choice) }], components: [row] });
            }
        } else {
            await getChoices.bind(this)(player1, player2, message).then(v => {
                player1Choice = v.p1choice;
                player2Choice = v.p2choice;

                if (no !== false) return sent.edit({ components: [], embeds: [{ color: this.colors.errorEmbed, title: `I was unable to DM ${no}, so please open DM than try again.` }] })

                let row = new Discord.MessageActionRow().addComponents(new Discord.MessageButton().setCustomId("y7ghjuiojioujoj").setDisabled(true).setStyle("SUCCESS").setEmoji("ðŸ•Š").setLabel("Game Ended"))

                if (player1Choice === player2Choice) { // draw
                    sent.edit({ embeds: [{ color: this.colors.drawEmbed, title: this.drawEndTitle, description: this.drawEndDescription.replace(/{player1}/g, message.author.username).replace(/{player1move}/g, player1Choice).replace(/{player2}/g, player2.username).replace(/{player2move}/g, player2Choice) }], components: [row] });
                } else if ((player1Choice === "âœŠ" && player2Choice === "âœŒï¸") || (player1Choice === "ðŸ¤š" && player2Choice === "âœŠ") || (player1Choice === "âœŒï¸" && player2Choice === "ðŸ¤š")) { // player 1 won
                    sent.edit({ embeds: [{ color: this.colors.endEmbed, title: this.endTitle.replace(/{winner}/g, message.author.username).replace(/{looser}/g, player2.username), description: this.endDescription.replace(/{winner}/g, message.author.username).replace(/{winnermove}/g, player1Choice).replace(/{looser}/g, player2.username || "Bot").replace(/{loosermove}/g, player2Choice) }], components: [row] });
                } else { // player 2 won
                    sent.edit({ embeds: [{ color: this.colors.endEmbed, title: this.endTitle.replace(/{looser}/g, message.author.username).replace(/{winner}/g, player2.username), description: this.endDescription.replace(/{looser}/g, message.author.username).replace(/{loosermove}/g, player1Choice).replace(/{winner}/g, player2.username || "Bot").replace(/{winnermove}/g, player2Choice) }], components: [row] });
                }
            }).catch(e => {
                no === true;
            });
        }
    }
}

module.exports = rps;

/**
 * @typedef {Object} Colors The colors to use in embeds.
 * @property {String} readyEmbed The color for ready embed.
 * @property {String} errorEmbed The color for error embed.
 * @property {String} drawEmbed The color for draw embed.
 * @property {String} endEmbed The color for draw embed.
 */
