// discord-tictactoe/dist/src/tictactoe/Game.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Player_1 = require("./Player");
class Game {
    constructor(boardSize) {
        this._boardSize = boardSize;
        this._board = [];
        this._winner = 0;
        this._currentPlayer = Math.random() < 0.5 ? 1 : 2;
        for (let i = 0; i < boardSize * boardSize; i++) {
            this._board[i] = 0;
        }
    }
    get boardSize() {
        return this._boardSize;
    }
    get board() {
        return this._board;
    }
    get finished() {
        return this.winner !== 0 || this.boardFull;
    }
    get currentPlayer() {
        return this._currentPlayer;
    }
    get winner() {
        return this._winner;
    }
    get boardFull() {
        return this.board.every(cell => cell !== 0);
    }
    get boardEmpty() {
        return this.board.every(cell => cell === 0);
    }
    get emptyCellAmount() {
        return this.board.filter(cell => cell === 0).length;
    }
    clone() {
        console.log("cloned");
        const game = new Game(this.boardSize);
        for (let i = 0; i < this.board.length; i++) {
            game.board[i] = this.board[i];
        }
        return game;
    }
    isMoveValid(position) {
        return position < this.board.length && this.board[position] === 0;
    }
    updateBoard(player, position) {
        this.board[position] = player;
        this._winner = this.computeWinner();
    }
    nextPlayer() {
        this._currentPlayer = (0, Player_1.getOpponent)(this.currentPlayer);
    }
    computeWinner() {
        var isFound = false;
        var values = new Array(this.boardSize);
        //compute winner by populating each row and column, then checking for rows
        for (let i = 0; i < this.boardSize; i++) {
            //set isFound to true and then see if it is made false during the check
            isFound = true;
            for (let j = 0; j < this.boardSize; j++) {
                values[j] = this.toIndex(i, j);
            }

            //The check to determine if winner has been found
            for (let j = 0; j < this.boardSize - 1; j++) {
                if (!this.validEquals(values[j], values[j + 1])) {
                    isFound = false;
                    break;
                }
            }

            if (isFound) {
                return this.board[values[0]];
            }
        }

        for (let i = 0; i < this.boardSize; i++) {
            //set isFound to true and then see if it is made false during the check
            isFound = true;
            for (let j = 0; j < this.boardSize; j++) {
                values[j] = this.toIndex(j, i);
            }

            //The check to determine if winner has been found
            for (let j = 0; j < this.boardSize - 1; j++) {
                if (!this.validEquals(values[j], values[j + 1])) {
                    isFound = false;
                    break;
                }
            }

            if (isFound) {
                return this.board[values[0]];
            }
        }

        //top left to bottom right check
        isFound = true;
        for (let i = 0; i < this.boardSize; i++) {
            values[i] = this.toIndex(i, i);
        }

        for (let i = 0; i < this.boardSize - 1; i++) {
            if (!this.validEquals(values[i], values[i + 1])) {
                isFound = false;
                break;
            }
        }
        if (isFound) {
            return this.board[values[0]];
        }

        //top right to bottom left check
        isFound = true;
        for (let i = 0; i < this.boardSize; i++) {
            values[i] = this.toIndex(i, this.boardSize-1-i);
        }

        for (let i = 0; i < this.boardSize - 1; i++) {
            if (!this.validEquals(values[i], values[i + 1])) {
                isFound = false;
                break;
            }
        }
        if (isFound) {
            return this.board[values[0]];
        }
        return 0;
    }
    toIndex(row, column) {
        return row * this.boardSize + column;
    }
    validEquals(position1, position2) {
        return (this.board[position1] !== 0 && this.board[position1] === this.board[position2]);
    }
}
exports.default = Game;
